### PR TITLE
memory leaks: don't pass process stack via context

### PR DESCRIPTION
**Pull request by ltalirz**
workaround for #4698 

Fix remaining reference to slurm calculation in #4603 

Note: This may require python >= 3.7, but we anyhow dropped python 3.6 support

### CODE REVIEW

File: aiida/engine/transports.py [hunk 1]
```diff
@@ -96,7 +97,7 @@ def do_open():
                         transport_request.future.set_result(transport)
 
             # Save the handle so that we can cancel the callback if the user no longer wants it
-            open_callback_handle = self._loop.call_later(safe_open_interval, do_open)
+            open_callback_handle = self._loop.call_later(safe_open_interval, do_open, context=contextvars.Context())
```
Comment (chrisjsewell) [inline aiida/engine/transports.py:100]: Add a comment here about why it is (currently) necessary to set a specific context

**Review by chrisjsewell — CHANGES_REQUESTED** [review]
Shall we devise a unit test now, to ensure there are no references remaining to the process

File: aiida/engine/transports.py (hunk 1 above)
Comment (chrisjsewell) [reply aiida/engine/transports.py:100]: it also appears that context is not part of the `AbstractEventLoop` 🤔 :

> aiida/engine/transports.py:100: error: Unexpected keyword argument "context" for "call_later" of "AbstractEventLoop"  [call-arg]

you will need to add `type: ignore[call-arg]`

**Comment by chrisjsewell** [comment]
As discussed with @ltalirz on Slack, using https://github.com/aiidateam/aiida-integration-tests, in `verdi shell` no processes are now left referenced:

```python
In [1]: import gc 
   ...:  
   ...: from pympler import muppy, refbrowser 
   ...:  
   ...: from aiida_sleep.cli import run_calc 
   ...: from aiida_sleep.sleep_job import SleepCalculation 
   ...:  
   ...: run_calc( 
   ...:     code="sleep@slurm", payload=int(5e5), output_dict=int(5e5), output_array=int(5e5) 
   ...: )                                                                                                          
Out[1]: <CalcJobNode: uuid: 12474cd9-00ae-4777-8dfc-726b640e45dd (pk: 13) (aiida.calculations:sleep)>

In [2]: gc.collect() 
   ...: all_objects = muppy.get_objects() 
   ...: calcs = [ 
   ...:     o 
   ...:     for o in all_objects 
   ...:     if hasattr(o, "__class__") and isinstance(o, SleepCalculation) 
   ...: ] 
   ...: len(calcs)                                                                                                 
Out[2]: 0
```

It is of note that, before garbage collection the memory usage is ~700MiB, then after garbage collection it is reduced to ~450MiB (before the PR it remained ~700MiB).
The garbage collection does not appear to automatically trigger, most probably because there are still some circular references (see https://github.com/aiidateam/plumpy/issues/197 and https://github.com/aiidateam/plumpy/issues/198) which can be garbage collected by Python but not automatically, from http://docs.python.org/reference/datamodel.html:

> CPython implementation detail: CPython currently uses a reference-counting scheme with (optional) delayed detection of cyclically linked garbage, which collects most objects as soon as they become unreachable, but is not guaranteed to collect garbage containing circular references. See the documentation of the gc module for information on controlling the collection of cyclic garbage. Other implementations act differently and CPython may change. Do not depend on immediate finalization of objects when they become unreachable (ex: always close files).

This manifests in calculations submitted to the daemon, which this PR does not change the memory:

`aiida-sleep calc -o 500000 -p 500000 -a 500000 -c sleep@local --submit` => ~800MiB

**Comment by chrisjsewell** [comment]
> It would be assumed that on garbage collection this number would be reduced.

I can confirm that after adding code to "fix" aiidateam/plumpy#197 and aiidateam/plumpy#198, the final memory is greatly reduced:

`aiida-sleep calc -o 500000 -p 500000 -a 500000 -c sleep@slurm --submit` => 470.6MiB

**Comment by ltalirz** [comment]
Just to be safe: the test does catch process instances that remain in memory 
https://github.com/aiidateam/aiida-core/pull/4699/checks?check_run_id=1808258216#step:9:91

I believe these stem from local variables; will refactor now (and add a sys.exit unless no remaining references are found)

**Comment by ltalirz** [comment]
I could also add an if/else depending on python version.

plumpy is using [aiocontextvars](https://github.com/fantix/aiocontextvars#compatibility) for python 3.6, and if I understand correctly, then we can simply omit the `context=` fix there since this package does *not* copy the context automatically.

**Comment by csadorf** [comment]
> I could also add an if/else depending on python version.
> 
> plumpy is using [aiocontextvars](https://github.com/fantix/aiocontextvars#compatibility) for python 3.6, and if I understand correctly, then we can simply omit the `context=` fix there since this package does *not* copy the context automatically.
> 
> @chrisjsewell @csadorf What do we do?
> Shall I drop python3.6 from CI now or do I include a fix?

I was going to drop it during the coding week, but if this is blocking, I suggest you just drop it now.

**Comment by ltalirz** [comment]
> I was going to drop it during the coding week, but if this is blocking, I suggest you just drop it now.

Opened https://github.com/aiidateam/aiida-core/pull/4701

**Comment by ltalirz** [comment]
@chrisjsewell Even after refactoring the daemon test I'm not able to get rid of all process instances in memory.

Do you spot where the two references may be coming from?

There's a couple more places using `call_soon` and `call_later`, I'm adding some more "empty contexts" now but in some places I guess the context is needed.

**Comment by chrisjsewell** [comment]
> Even after refactoring the daemon test I'm not able to get rid of all process instances in memory.

"This tests only processes that were `run` in the same interpreter, not those that were `submitted`"

Given that this not does not even test the daemon, I don't really see why this test should go in `.ci/test_daemon.py` and not in a more targeted unit test in the core `tests/` folder?

**Comment by ltalirz** [comment]
> Given that this not does not even test the daemon, I don't really see why this test should go in `.ci/test_daemon.py` and not in a more targeted unit test in the core `tests/` folder?

Because I want to test the ssh transport

**Comment by chrisjsewell** [comment]
> Because I want to test the ssh transport.

Maybe I'm missing something, but I don't see any computers used with an ssh transport?

**Comment by ltalirz** [comment]
Ha, you're right :-D

I just remembered that we used to have tests against some ssh docker container and thought that was part of the setup.

Does this mean the ssh transport is currently untested?

It would definitely be nice to test it, since this was where we originally discovered the issue.

**Comment by chrisjsewell** [comment]
> Edit: seems at least relatively well covered by the unit tests.

Yeh you can set up an ssh transport (and I assume computer) that just connects locally: https://github.com/aiidateam/aiida-core/blob/48fa47584c993cdac019c6199bc045a4c19da152/tests/transports/test_ssh.py#L62

> But is there also an integration test?

But no I don't see any tests of calculations that use such a computer

**Comment by chrisjsewell** [comment]
> I guess that will become easier with the molecule setup

Thats the idea

**Comment by ltalirz** [comment]
> But no I don't see any tests of calculations that use such a computer

There used to be https://github.com/aiidateam/aiida-core/blob/v1.0.0/.travis.yml

**Comment by chrisjsewell** [comment]
I think lets start simple, and just try to create a very basic unit test, that runs a single calcjob on a local/direct computer and checks for references

**Comment by ltalirz** [comment]
Just for the record, while the test runs fine when I run it using
```
pytest tests/engine/test_run.py::TestRun::test_run
```

[there are loads of processes in memory when it is run as part of the test suite](https://github.com/aiidateam/aiida-core/pull/4699/checks?check_run_id=1854470422#step:9:435):
```
E       AssertionError: Memory leak: process instances remain in memory: [<aiida.tools.data.cif._get_aiida_structure_ase_inline object at 0x7f80f0a96280>, <aiida.tools.data.cif._get_aiida_structure_ase_inline object at 0x7f80f0b58d90>, <aiida.tools.data.cif._get_aiida_structure_pymatgen_inline object at 0x7f80f0adfd90>, <aiida.tools.data.cif.refine_inline object at 0x7f80ee5ed9d0>, <aiida.tools.data.cif.refine_inline object at 0x7f80ee5bf4c0>, <aiida.tools.data.array.trajectory._get_aiida_structure_inline object at 0x7f80ee1badf0>, <tests.cmdline.commands.test_data.connect_structure_bands object at 0x7f80f0f02940>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80edeac040>, <tests.engine.test_calcfunctions.add_calcfunction object at 0x7f80e7603a90>, <tests.engine.test_calcfunctions.return_stored_calcfunction object at 0x7f80e7498040>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80e7760be0>, <tests.engine.test_process.ProcessStackTest object at 0x7f80edcc7130>, <tests.engine.test_process_function.function_excepts object at 0x7f80e671e370>, <tests.engine.test_calcfunctions.execution_counter_calcfunction object at 0x7f80ee59e670>, <tests.engine.test_process_function.function_out_unstored object at 0x7f80e6b54e80>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80e760de50>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ee0c36d0>, <tests.engine.test_process.TestProcess.test_output_dictionary.<locals>.TestProcess1 object at 0x7f80e6e9aa90>, <tests.engine.test_process.TestProcess.test_output_validation_error.<locals>.TestProcess1 object at 0x7f80e72aca90>, <tests.engine.test_process_function.function_exit_code object at 0x7f80e6610490>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ed96cc70>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ed700550>, <__main__.wf object at 0x7f80f0c5c250>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ed8173a0>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80edf7ce50>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ee418d60>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ee418c70>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80f0dd6520>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ed65ed60>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ed5a13a0>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ecec5a60>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ec24cca0>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ec1e8e80>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ec24c880>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ec1e8d90>, <tests.engine.test_calc_job.FileCalcJob object at 0x7f80ec121b80>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ec237eb0>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ec102040>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ec08fee0>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ec070670>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ec2371c0>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ec08f7c0>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ec0a6280>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80e77549a0>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80e77666d0>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80e76c9c10>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80ec043d00>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80e7699fd0>, <tests.engine.test_calcfunctions.execution_counter_calcfunction object at 0x7f80e77a76a0>, <tests.engine.test_calcfunctions.add_calcfunction object at 0x7f80e774b160>, <tests.engine.test_calcfunctions.add_calcfunction object at 0x7f80ec0fdb20>, <tests.engine.test_calcfunctions.add_calcfunction object at 0x7f80e7644310>, <tests.engine.test_calcfunctions.add_calcfunction object at 0x7f80edaeb490>, <tests.engine.test_calcfunctions.add_calcfunction object at 0x7f80e7709a30>, <tests.engine.test_calcfunctions.add_calcfunction object at 0x7f80e74f6a00>, <tests.engine.test_calcfunctions.test_calcfunction_caller object at 0x7f80e75235b0>, <tests.engine.test_calcfunctions.add_calcfunction object at 0x7f80e73d47f0>, <tests.engine.test_calcfunctions.test_calcfunction_caller object at 0x7f80e7416eb0>, <tests.utils.processes.DummyProcess object at 0x7f80e74945b0>, <tests.engine.test_launch.add object at 0x7f80e73cb760>, <tests.engine.test_launch.add object at 0x7f80ec19faf0>, <tests.engine.test_launch.add object at 0x7f80e7347e80>, <tests.engine.test_launch.AddWorkChain object at 0x7f80e725a310>, <tests.engine.test_launch.AddWorkChain object at 0x7f80e72cd700>, <tests.engine.test_launch.AddWorkChain object at 0x7f80e73a4ca0>, <tests.engine.test_launch.AddWorkChain object at 0x7f80e7384f10>, <tests.engine.test_launch.AddWorkChain object at 0x7f80e72acc10>, <tests.engine.test_launch.AddWorkChain object at 0x7f80e72ac760>, <tests.engine.test_launch.FileCalcJob object at 0x7f80e7201490>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80e717d850>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80e7197850>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80e7105a90>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80f4733dc0>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80e7130c70>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80e7130340>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80e709b3d0>, <aiida.calculations.arithmetic.add.ArithmeticAddCalculation object at 0x7f80e716b0a0>, <tests.utils.processes.DummyProcess object at 0x7f80e71656d0>, <tests.utils.processes.DummyProcess object at 0x7f80e71a3130>, <tests.utils.processes.DummyProcess object at 0x7f80e6fc47c0>, <tests.utils.processes.InvalidateCaching object at 0x7f80e6fa15b0>, <tests.utils.processes.InvalidateCaching object at 0x7f80e703a400>, <tests.utils.processes.InvalidateCaching object at 0x7f80e6f40e20>, <tests.utils.processes.InvalidateCaching object at 0x7f80e6ff8220>, <tests.utils.processes.DummyProcess object at 0x7f80e6f446a0>, <tests.utils.processes.BadOutput object at 0x7f80e6ebd490>, <tests.utils.processes.DummyProcess object at 0x7f80e71f06a0>, <tests.engine.test_process.TestProcess.test_output_validation_error.<locals>.TestProcess1 object at 0x7f80e6fee5e0>, <aiida.calculations.templatereplacer.TemplatereplacerCalculation object at 0x7f80e6e57b50>, <tests.utils.processes.DummyProcess object at 0x7f80e6e9af10>, <tests.utils.processes.DummyProcess object at 0x7f80e72017c0>, <tests.utils.processes.IsValidCacheHook object at 0x7f80e6efd820>, <tests.utils.processes.IsValidCacheHook object at 0x7f80e6f19d00>, <tests.utils.processes.IsValidCacheHook object at 0x7f80e6ddc1c0>, <tests.utils.processes.IsValidCacheHook object at 0x7f80e6d40b20>, <tests.utils.processes.DummyProcess object at 0x7f80e6f19c40>, <tests.engine.test_process_function.function_args_with_default object at 0x7f80e6ee11c0>, <tests.engine.test_process_function.function_args object at 0x7f80e6c016d0>, <tests.engine.test_process_function.function_args_and_kwargs object at 0x7f80e6cea8e0>, <tests.engine.test_process_function.function_args_and_kwargs object at 0x7f80e6cebaf0>, <tests.engine.test_process_function.function_args_and_default object at 0x7f80e6b44b50>, <tests.engine.test_process_function.function_args_and_default object at 0x7f80e6befb20>, <tests.engine.test_process_function.function_args_with_default object at 0x7f80e6b07a30>, <tests.engine.test_process_function.function_args_with_default object at 0x7f80e6cce7f0>, <tests.engine.test_process_function.function_default_label object at 0x7f80e6b19f10>, <tests.engine.test_process_function.function_defaults object at 0x7f80e6c3a8b0>, <tests.engine.test_process_function.function_default_label object at 0x7f80e6be53d0>, <tests.engine.test_process_function.function_defaults object at 0x7f80e6ae1cd0>, <tests.engine.test_process_function.function_kwargs object at 0x7f80e6b0baf0>, <tests.engine.test_process_function.function_kwargs object at 0x7f80e6b0b040>, <tests.engine.test_process_function.function_args object at 0x7f80e6a351c0>, <tests.engine.test_process_function.function_args_with_default object at 0x7f80e699fb50>, <tests.engine.test_process_function.function_kwargs object at 0x7f80e69b0160>, <tests.engine.test_process_function.function_args_and_default object at 0x7f80e692e5b0>, <tests.engine.test_process_function.function_args_and_kwargs object at 0x7f80e697c3a0>, <tests.engine.test_process_function.function_with_none_default object at 0x7f80e688d4f0>, <tests.engine.test_process_function.function_with_none_default object at 0x7f80e6941670>, <tests.engine.test_process_function.function_return_input object at 0x7f80e69f8ac0>, <tests.engine.test_process_function.function_return_input object at 0x7f80e6878790>, <tests.engine.test_process_function.function_return_input object at 0x7f80e6962880>, <tests.engine.test_process_function.function_return_input object at 0x7f80e695b2e0>, <tests.engine.test_process_function.function_return_true object at 0x7f80e67ef970>, <tests.engine.test_process_function.function_return_true object at 0x7f80e68bafa0>, <aiida.workflows.arithmetic.add_multiply.add_multiply object at 0x7f80e683ea30>, <tests.engine.test_process_function.function_args_with_default object at 0x7f80e68349a0>, <tests.engine.test_process_function.function_args_with_default object at 0x7f80e66f36a0>, <tests.engine.test_process_function.function_defaults object at 0x7f80e66c2c70>, <tests.engine.test_process_function.add_mul_wf object at 0x7f80e6d39a60>, <tests.engine.test_process_function.add object at 0x7f80e6f18700>, <tests.engine.test_process_function.mul object at 0x7f80e6d26e80>, <tests.engine.test_process_function.test_process_function object at 0x7f80e6516d60>, <tests.utils.processes.WaitProcess object at 0x7f80e650b5e0>, <tests.utils.processes.AddProcess object at 0x7f80e658e2e0>, <tests.utils.processes.WaitProcess object at 0x7f80e6633c70>, <tests.utils.processes.WaitProcess object at 0x7f80e654d970>, <tests.utils.processes.DummyProcess object at 0x7f80e64860a0>, <tests.utils.processes.DummyProcess object at 0x7f80e65b8310>]
```

File: aiida/engine/processes/calcjobs/manager.py [hunk 2]
```diff
@@ -180,15 +181,21 @@ async def updating():
             # Any outstanding requests?
             if self._update_requests_outstanding():
                 self._update_handle = self._loop.call_later(
-                    self._get_next_update_delay(), asyncio.ensure_future, updating()
+                    self._get_next_update_delay(),
+                    asyncio.ensure_future,
+                    updating(),
+                    context=contextvars.Context(),
```
Comment (ltalirz) [inline aiida/engine/processes/calcjobs/manager.py:187]: Note for context: these changes may not be necessary to get rid of the process reference - on the other side it seems they don't break anything (it's just checking whether the jobinfo is being updated)

**Comment by ltalirz** [comment]
I've added the ssh test in https://github.com/ltalirz/aiida-core/tree/add-ssh-test

Also there the tests pass (but it involves a little bit of changes in the setup) - I can include it here or I can add it in a separate PR - up to you @chrisjsewell

File: aiida/engine/processes/calcjobs/manager.py (hunk 2 above)
Comment (chrisjsewell) [reply aiida/engine/processes/calcjobs/manager.py]: yep that makes sense

File: .ci/pytest/test_memory_leaks.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
+# -*- coding: utf-8 -*-
+###########################################################################
+# Copyright (c), The AiiDA team. All rights reserved.                     #
+# This file is part of the AiiDA code.                                    #
+#                                                                         #
+# The code is hosted on GitHub at https://github.com/aiidateam/aiida-core #
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+"""Utilities for testing memory leakage."""
```
Comment (ltalirz) [inline .ci/pytest/test_memory_leaks.py:10]:
```suggestion
"""Testing against memory leakage of processes when running them in the python interpreter."""
```

**Comment by chrisjsewell** [comment]
> I've added the ssh test in https://github.com/ltalirz/aiida-core/tree/add-ssh-test
> Also there the tests pass (but it involves a little bit of changes in the setup) - I can include it here or I can add it in a separate PR - up to you @chrisjsewell

Yeh looking good 👍 , ermm I will leave that up to you. Will just get #4565 merged first as there will likely be some merge conflicts to sort out here after that

**Review by chrisjsewell — CHANGES_REQUESTED** [review]
just the conflict to resolve, then good to go 👍

**Review by chrisjsewell — APPROVED** [review]

**Pull request merged**